### ISSUE TITLE
Windows install fails 0.12.12

### ISSUE THREAD

**Issue by edporteous** [bokeh/bokeh#7283]
I'm getting an error installing bokeh 0.12.12. Previous versions install okay on my PC. I did a fresh install of Python 3.6.3 and deleted the pip cache directory first.

Windows 10 32-bit
Python 3.6.3 32-bit

Running pip install yields:

```
> pip install bokeh==0.12.12
Collecting bokeh==0.12.12
  Using cached bokeh-0.12.12.tar.gz
Exception:
Traceback (most recent call last):
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\tarfile.py", line 2268, in utime
    os.utime(targetpath, (tarinfo.mtime, tarinfo.mtime))
OSError: [WinError 87] The parameter is incorrect

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\basecommand.py", line 215, in main
    status = self.run(options, args)
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\commands\install.py", line 324, in run
    requirement_set.prepare_files(finder)
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\req\req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\req\req_set.py", line 620, in _prepare_file
    session=self.session, hashes=hashes)
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\download.py", line 821, in unpack_url
    hashes=hashes
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\download.py", line 663, in unpack_http_url
    unpack_file(from_path, location, content_type, link)
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\utils\__init__.py", line 605, in unpack_file
    untar_file(filename, location)
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\site-packages\pip\utils\__init__.py", line 581, in untar_file
    tar.utime(member, path)
  File "k:\users\ed\appdata\local\programs\python\python36-32\lib\tarfile.py", line 2270, in utime
    raise ExtractError("could not change modification time")
tarfile.ExtractError: could not change modification time
```

**Comment by bryevdv** [comment]
ping @mattpap This was originally reported by @jakirkham as a problem with condo-build building the packages on condo-forge:

> Appears that conda-build is failing on Windows due to a missing aux.d.ts file during tarball extraction. There is no aux.d.ts in the repo, but there is an aux.ts. Not sure why the issue is absent on Linux. Traceback and link to the log included below.

Any ideas?

Until someone decides to help by volunteering to help with windows support and maintenance I guess we are just doomed to these possibilities. I can't take on more work.

**Comment by bryevdv** [comment]
My first inclination is that these files should not be in the sdist at all, so let me try updating the manifest to exclude them.

**Comment by bryevdv** [comment]
I've been trying to work on one thing for two weeks and still have not gotten a chance to even start it.

**Comment by bryevdv** [comment]
@edporteous If I push PRs do you have the ability to build sdists and test them on windows? We desperately need help with windows maintenance.

**Comment by edporteous** [comment]
I'm not an expert, but I'll give it a go.

**Comment by bryevdv** [comment]
@edporteous brilliant, thanks, I will have a PR shortly

**Comment by bryevdv** [comment]
@edporteous thinking a bit more, I am guessing since you reported a problem re: pip install, that means you don't use Anaconda/conda? In that case getting set up to build Bokeh from source will be much more laborious. I will try to get a VM up and running.

**Comment by edporteous** [comment]
Yes, I've stopped using conda since switching to Python 3.6.

**Comment by bryevdv** [comment]
@edporteous unrelated, would you mind sharing what about Python 3.6 prompted a switch from Anaconda?

**Comment by bryevdv** [comment]
I've merged the PR. I will build `0.12.13rc1` now which will hopefully facilitate easier testing by others

**Comment by edporteous** [comment]
@bryevdv Simply because I could never get numpy, etc. to install with pip on Python 2.7, but with Python 3.6 these libraries installs fine.

**Comment by bryevdv** [comment]
@jakirkham @edporteous The RC1 seems to pip install (from anaconda.org, we don't publish pre-releaes to pypi). Here is a screenshot of `0.12.12` failing and `0.12.13rc1` succeeding:

<img width="920" alt="screen shot 2017-12-06 at 10 20 08" src="https://user-images.githubusercontent.com/1078448/33672190-3f676084-da6f-11e7-8e98-3c7d32c60957.png">

can you both try pip installing using the instructions here:

https://bokeh.pydata.org/en/dev/docs/installation.html#developer-builds

**Comment by edporteous** [comment]
Fantastic! That works for me too.

**Comment by bryevdv** [comment]
OK after a few more confirmations I will kick off a full release build

**Comment by jakirkham** [comment]
Sure. Put up PR ( https://github.com/conda-forge/bokeh-feedstock/pull/17 ) to test building in `conda-forge`. Will report back with results.

**Comment by bryevdv** [comment]
There's also an issue with the conda package (only on windows) 
<img width="690" alt="screen shot 2017-12-06 at 11 37 05" src="https://user-images.githubusercontent.com/1078448/33675917-cfc546a0-da79-11e7-8a67-947e0f72146e.png">

Ray reports:

> file or directory: 'C:\\Users\\builder\\m64\\pkgs\\bokeh-0.12.13rc1-py_0\\site-packages\\bokeh\\server\\static\\js\\tree\\core\\aux.js.map'

Which is suspiciously related to the same `aux.d.ts` file I think. It's possibly an issue that the file is zero bytes long @mattpap what is going on?

**Comment by mattpap** [comment]
Neither of `aux.{js, js.map, d.ts}` should be empty (at least in this case, `*.d.ts` can be if there are no new public types). `*.js.map` is related only to `*.js`. All this shouldn't really matter, because `tree/` and (now) `types/` are excluded from build (right?).

**Comment by mingwandroid** [comment]
I am seeing:

```
import tarfile
tf = tarfile.open('C:/Users/builder/m64/pkgs/bokeh-0.12.13rc1-py_0.tar.bz2')
tf.extractall()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\builder\m64\lib\tarfile.py", line 2007, in extractall
    numeric_owner=numeric_owner)
  File "C:\Users\builder\m64\lib\tarfile.py", line 2049, in extract
    numeric_owner=numeric_owner)
  File "C:\Users\builder\m64\lib\tarfile.py", line 2119, in _extract_member
    self.makefile(tarinfo, targetpath)
  File "C:\Users\builder\m64\lib\tarfile.py", line 2160, in makefile
    with bltn_open(targetpath, "wb") as target:
FileNotFoundError: [Errno 2] No such file or directory: '.\\site-packages\\bokeh\\server\\static\\js\\tree\\core\\aux.js.map'
```

**Comment by bryevdv** [comment]
I looked at the locally built versions of these files and they are not empty. They seem completely unremarkable.

**Comment by mingwandroid** [comment]
they aren't empty in the tarfile either, just in some way not working right.

**Comment by bryevdv** [comment]
Here is `aux.d.ts`:
```
export interface Class<T> {
    new (...args: any[]): T;
    prototype: T;
}
export declare type Constructor<T = {}> = new (...args: any[]) => T;
```
And here is `aux.js.map`:
```
{"version":3,"sources":[],"names":[],"mappings":"","file":"aux.js","sourcesContent":[]}
```

**Comment by mattpap** [comment]
Both are correct.

**Comment by bryevdv** [comment]
does cygwin or mingw command line tar work?

**Comment by mingwandroid** [comment]
It seems the `core` directory is read-only.

**Comment by bryevdv** [comment]
In my local source tree that dir is 755 just like `server/static/js`

**Comment by mingwandroid** [comment]
I'll have to bring procmon to bear on this one I think.

**Comment by bryevdv** [comment]
@mingwandroid I appreciate your help, I do not have the experience to figure this out. In the mean time I am gong to pull the packaged from pypi and anaconda.org

**Comment by mingwandroid** [comment]
ok, I think that is sensible.

**Comment by bryevdv** [comment]
The `0.12.13rc1` packages are still available in the `dev` channel if you need them

**Comment by mingwandroid** [comment]
At a low level, we see:
```
12:05:28.8186126 PM	python.exe	6044	CreateFile	\\vboxsrv\conda\AGGREGATE\SITE-PACKAGES\bokeh\server\static\js\tree\core	SUCCESS	Desired Access: Read Attributes, Synchronize, Disposition: Open, Options: Synchronous IO Non-Alert, Open Reparse Point, Attributes: N, ShareMode: None, AllocationSize: n/a, OpenResult: Opened
```
(which is fine and common for all directories created while untarring this package)

but then very soon after we see:
```
12:05:28.8203819 PM	python.exe	6044	CreateFile	\\vboxsrv\conda\AGGREGATE\SITE-PACKAGES\bokeh\server\static\js\tree\core	SUCCESS	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: Opened
```
which is not fine. It creates a symlink basically to, well, I don't know what.

**Comment by mattpap** [comment]
On Windows I neither can't merge most recent master, nor clone (or checkout to be specific) the repository. In both cases git fails on `core/aux.ts`.

**Comment by mingwandroid** [comment]
OK, I think I know what it is.

Windows has very low level hacks regarding files that are called 'aux' (and 'prn' and some others).

**Comment by bryevdv** [comment]
unbelievable.

**Comment by mingwandroid** [comment]
Indeed. https://github.com/BeyondTheClouds/VMPlaceS/issues/10

**Comment by mingwandroid** [comment]
and: https://blogs.msdn.microsoft.com/oldnewthing/20031022-00/?p=42073

**Comment by jakirkham** [comment]
Impressive debugging, @mingwandroid. So the recommendation is rename the file then I guess?

**Comment by bryevdv** [comment]
Here is a list:

https://msdn.microsoft.com/en-us/library/aa578688.aspx

**Comment by bryevdv** [comment]
I will make a new PR to rename the file and add a code quality check to prevent any files with those names, best to do until I can learn enough about appveyor to get it up and running.

**Comment by mattpap** [comment]
I understand backwards compatibility, but at least they could have provided more meaningful error message than "No such file or directory". This could have been resolved in seconds.

**Comment by mingwandroid** [comment]
conda-build could do with such a check I think, both for Windows packages and all noarch packages too, what do you think @msarahan?

**Comment by msarahan** [comment]
I'd rather put any checking into conda-verify, but yes, I agree that this is annoying enough that we could help save people some time by adding a check for it.

**Comment by jakirkham** [comment]
When would the check happen? It sounds like as long as the `aux*` file appears a Windows user can't even use the repo based on @mattpap's experience. Also when we tried to build this in conda-forge we weren't able to extract the tarball. So this was an issue prepackage creation. Is there some sort of source check that `conda-verify` can run?

**Comment by mingwandroid** [comment]
When making noarch packages it can be done of course, but for other things, good point.

**Comment by bryevdv** [comment]
Not sure what you guys will do, but I am doing this for Bokeh:

* (now) a code quality to check to check any file in the repo under `bokeh` has a filename root from that list
* (later) an appveyor test to exercise package installation 

the code quality check is not perfect but should avoid getting anything into github with a bad name, and will check any artifacts built that happen to be in the tree as well.

**Comment by jakirkham** [comment]
> When making noarch packages it can be done of course, but for other things, good point.

Would add the caveat when making noarch packages not on Windows. Otherwise yes agreed it is still valuable to have this check for noarch packages either during the build or on the final product.

**Comment by jakirkham** [comment]
If it helps @bryevdv, feel free to take an AppVeyor file from conda-forge and tweak it to your needs. These are pretty highly tuned, so make for a great starting point.

**Comment by bryevdv** [comment]
@jakirkham is the delay for your build because C-F is a large org? Or is that delay typical of any free OSS project?

**Comment by jakirkham** [comment]
The delay is because of the large org.

AppVeyor typically provides 1 build with some delay of VM startup (seconds) on a low end machine. This is pretty reasonable for an org with a repo or so on the service.

conda-forge has a nicer tier on AppVeyor so gets 2 builds with nearly no delay on a high end machine, but the number of repos takes a toll. Hence the largeish queue. We do some things to cut it down like canceling old builds and the like. This is all included in the AppVeyor configuration file.

There are paid tiers on AppVeyor with everything between free and what conda-forge has (plus more beyond).

**Comment by bryevdv** [comment]
`0.12.13rc2` is building

**Comment by bryevdv** [comment]
@jakirkham @mattpap @edporteous @mingwandroid I have tested installing `0.12.13rc2` with both pip and conda on a windows VM. I'd like to get one more confirmation before starting the release build.

**Comment by edporteous** [comment]
Installs fine for me on Windows 10 / Python 3.6 / pip.

**Comment by bryevdv** [comment]
@mingwandroid can you try as well, when you get a chance?

**Comment by bryevdv** [comment]
OK was hoping for another conda confirm but given that `rc1` exhibited the problem for me and `rc2` does not, I'm going to go ahead and kick off the release build.

**Comment by jakirkham** [comment]
FWIW `bokeh` version `0.12.13rc2` built fine in `conda-forge`.

**Comment by bryevdv** [comment]
@jakirkham is there are way to start the appveyor job programmatically (from TravisCI) or to have an independent appveyor job wait on some condition? The TravisCI build stage builds and archives an sdist and it's this sdist that I would like to have appveyor use for testing.

**Comment by jakirkham** [comment]
They have a REST API. Maybe [this]( https://www.appveyor.com/docs/api/projects-builds/#start-build-of-branch-most-recent-commit ) is what you are looking for?
